### PR TITLE
[API2.0] add dropout, dropout2d and dropout3d in nn.functional and nn.layer

### PR DESCRIPTION
**Pull request by huangjun12**
### PR types
New features

### PR changes
APIs

### Describe
Add apis of dropout:
(1)  paddle.nn.functional.dropout(x, p=0.5, axis=None, training=True, mode="upscale_in_train”, name=None) 
(2) paddle.nn.functional.dropout2d(x, p=0.5, training=True, name=None) 
(3) paddle.nn.functional.dropout3d(x, p=0.5, training=True, name=None)
(4) paddle.nn.Dropout(p=0.5, axis=None, mode="upscale_in_train”, name=None)
(5) paddle.nn.Dropout2d(p=0.5, data_format='NCHW', name=None)
(6) paddle.nn.Dropout3d(p=0.5, data_format='NCDHW', name=None)

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_dropout_op.py [hunk 1]
```diff
@@ -236,5 +237,392 @@ def test_dtype():
...
+class TestDropoutFAPI(unittest.TestCase):
+    def setUp(self):
+        np.random.seed(123)
+        self.places = [fluid.CPUPlace()]
+        if core.is_compiled_with_cuda():
+            self.places.append(fluid.CUDAPlace(0))
+
+    def check_static_result(self, place):
+        with fluid.program_guard(fluid.Program(), fluid.Program()):
+            input = fluid.data(name="input", shape=[4, 4], dtype="float32")
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/tests/unittests/test_dropout_op.py]: shape要求元素个数大于100

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -950,6 +947,8 @@ def dropout(x,
     probability) the outputs of some units to zero, while others are remain
     unchanged.
 
+    ***Warning*** This API is deprecated since 2.0, we recommend to use `paddle.nn.functional.dropout` to achieve the same function.
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/layers/nn.py]: 加一下@deprecated装饰器，参考https://github.com/PaddlePaddle/Paddle/pull/25912

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by iclementine — COMMENTED** [review]
是否叫 Dropout2D Dropout3D 更好？和 paddle 已有的一些 Conv2D Conv3D 之类的一致。

File: python/paddle/nn/functional/common.py [hunk 3]
```diff
@@ -446,3 +450,353 @@ def _is_list_or_turple_(data):
...
+            y_01 = paddle.nn.functional.dropout(x, axis=[0,1])
+            print(x.numpy())
+            print(y_0.numpy())
+            print(y_1.numpy())
+            print(y_01.numpy())
+
+    """
+    assert isinstance(p, (float, int)), "p argument should be a number"
+    assert 0 <= p <= 1, "p argument should between 0 and 1"
+    assert mode in (
```
Comment (iclementine) [inline python/paddle/nn/functional/common.py]: 这些地方建议用 “if 不满足条件，raise ValueError 或者其他 Error 的形式。”

File: python/paddle/fluid/tests/unittests/test_dropout_op.py (hunk 1 above)
Comment (huangjun12) [reply python/paddle/fluid/tests/unittests/test_dropout_op.py]: Done

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (huangjun12) [reply python/paddle/fluid/layers/nn.py]: Done

**Comment by huangjun12** [comment]
> 是否叫 Dropout2D Dropout3D 更好？和 paddle 已有的一些 Conv2D Conv3D 之类的一致。

Done

File: python/paddle/nn/functional/common.py (hunk 3 above)
Comment (huangjun12) [reply python/paddle/nn/functional/common.py]: Done

**Review by iclementine — DISMISSED** [review]
LGTM
不过一些 Tensor * scalar 的情形用 layers.scale 比用 multiply 会更高效一些

**Comment by huangjun12** [comment]
> LGTM
> 不过一些 Tensor * scalar 的情形用 layers.scale 比用 multiply 会更高效一些

Done

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/nn/functional/common.py [hunk 4]
```diff
@@ -456,6 +459,342 @@ def _is_list_or_turple_(data):
...
+    outputs of some units to zero, while upscale others according to the given
+    dropout probability.
+
+    Args:
+        x (Tensor): The input tensor. The data type is float32 or float64.
+        p (float | int): Probability of setting units to zero. Default 0.5.
+        axis (int | list): The axis along which the dropout is performed. Default None.
+        training (bool): A flag indicating whether it is in train phrase or not. Default True.
+        name (str, optional): Name for the operation (optional, default is None). For more information, please refer to :ref:`api_guide_Name`.
+        mode(str): ['upscale_in_train'(default) | 'downscale_in_infer']
```
Comment (jzhang533) [inline python/paddle/nn/functional/common.py:480]: 文档中顺序与api定义时顺序一致。下同

File: python/paddle/nn/layer/common.py [hunk 5]
```diff
@@ -348,6 +352,189 @@ def forward(self, input):
...
+class Dropout(layers.Layer):
+    """
+    Dropout is a regularization technique for reducing overfitting by preventing
+    neuron co-adaption during training as described in the paper:
+    `Improving neural networks by preventing co-adaptation of feature detectors <https://arxiv.org/abs/1207.0580>`_ 
+    The dropout operator randomly sets the outputs of some units to zero, while upscale others
+    according to the given dropout probability.
+
+    See ``paddle.nn.functional.dropout`` for more details.
+    In dygraph mode, please use ``eval()`` to indicate whether it is in test phrase or not.
```
Comment (jzhang533) [inline python/paddle/nn/layer/common.py:364]: 使用eval()就是确定转到测试阶段吧？
并不是`whether`

File: python/paddle/nn/layer/common.py [hunk 6]
```diff
@@ -348,6 +352,189 @@ def forward(self, input):
...
+            print(x.numpy())
+            print(y_train.numpy())
+            print(y_test.numpy())
+   """
+
+    def __init__(self, p=0.5, axis=None, mode="upscale_in_train", name=None):
+        super(Dropout, self).__init__()
+
+        self.p = p
+        self.training = _dygraph_tracer()._train_mode
```
Comment (jzhang533) [inline python/paddle/nn/layer/common.py:407]: 不需要这行？直接用Layer里的training

File: python/paddle/nn/layer/common.py [hunk 7]
```diff
@@ -348,6 +352,189 @@ def forward(self, input):
...
+            print(x.numpy())
+            print(y_train.numpy())
+            print(y_test.numpy())
+   """
+
+    def __init__(self, p=0.5, data_format='NCDHW', name=None):
+        super(Dropout3D, self).__init__()
+
+        self.p = p
+        self.training = _dygraph_tracer()._train_mode
```
Comment (jzhang533) [inline python/paddle/nn/layer/common.py:524]: 不需要这行？

**Review by jzhang533 — COMMENTED** [review]
另外，是否需要把dropout的实现拆到另一个文件里？而不是放到common.py里，防止conflict。

**Review by jzhang533 — APPROVED** [review]
will merge
some followup:
- doc fix discussed here
- separate the implementation in another file
- remove _dygraph_tracer()._train_mode

**Comment by huangjun12** [comment]
> will merge
> some followup:
> 
> * doc fix discussed here
> * separate the implementation in another file
> * remove _dygraph_tracer()._train_mode

to do in next pr, thx~
